exonum/src/runtime/error.rs

```diff
@@ -26,57 +26,112 @@ use crate::{
     proto::schema::runtime as runtime_proto,
 };
 
-/// Kind of execution error, indicates in which module error occurred.
+/// Kind of execution error, indicates the location of the error.
+///
+/// # Note to Runtime Developers
+///
+/// When should a runtime use different kinds of errors? Here's the guide.
+///
+/// ## `Service` errors
+///
+/// Use `Service` kind if the error has occurred in the service code and it makes sense to notify
+/// users about the error cause and/or its precise kind. These errors are generally raised
+/// if the input data (e.g., the transaction payload) violate certain invariants imposed by the service.
+/// For example, a `Service` error can be raised if the sender of a transfer transaction
+/// in the token service does not have sufficient amount of tokens.
+///
+/// ## `Unchecked` errors
+///
+/// Use `Unchecked` kind if the error has occurred in the service code, and:
+///
+/// - It is caused by the environment (e.g., out-of-memory)
+/// - It should never occur during normal execution (e.g., out-of-bounds indexing, null pointer)
+/// - It otherwise makes no sense to report to users in detail.
+///
+/// This kind of errors generally corresponds to panics in Rust and unchecked exceptions in Java.
+/// `Unchecked` errors are assumed to be reasonably rare by the framework; e.g., they are logged
+/// with a higher priority than other kinds.
+///
+/// Runtime environments can have mechanisms to convert `Unchecked` errors to `Service` ones
+/// (e.g., by catching exceptions in Java or calling [`catch_unwind`] in Rust),
+/// but whether it makes sense heavily depends on the use case.
+///
+/// ## `Dispatcher` errors
+///
+/// Use `Dispatcher` kind if the error has occurred while dispatching the request (i.e., *not*
+/// in the client code). See [`DispatcherError`] for more details.
+///
+/// ## `Runtime` errors
+///
+/// Use `Runtime` kind if the error has occurred in the runtime code and it makes sense to report
+/// the error to the users. A primary example here is artifact deployment: if deployment has failed,
+/// a `Runtime` error can provide more details about the cause.
+///
+/// ## Panics
```

Comment:
I believe the header level should be different and maybe renamed to "Policy on panics", since it's not really the `ErrorKind`.